samples/api/swapchain_recreation/swapchain_recreation.cpp

```diff
@@ -0,0 +1,950 @@
+
+	if (result != VK_SUCCESS && result != VK_SUBOPTIMAL_KHR)
+	{
+		// If failed, fence is untouched, recycle it.
+		//
+		// The semaphore is also untouched, but it may be used in the retry of
+		// vkAcquireNextImageKHR.  It is nevertheless cleaned up after cpu throttling
+		// automatically.
+		recycle_fence(acquire_fence);
+		return result;
+	}
+
+	associate_fence_with_present_history(*index, acquire_fence);
+
+	return result;
+}
+
+/**
+ * @brief Presents an image to the swapchain.
+ * @param index The swapchain index previously obtained from @ref acquire_next_image.
+ * @returns Vulkan result code
+ */
+VkResult SwapchainRecreation::present_image(uint32_t index)
+{
+	PerFrame &frame = submit_history[submit_history_index];
+
+	VkPresentInfoKHR present{VK_STRUCTURE_TYPE_PRESENT_INFO_KHR};
+	present.swapchainCount     = 1;
+	present.pSwapchains        = &swapchain;
+	present.pImageIndices      = &index;
+	present.waitSemaphoreCount = 1;
+	present.pWaitSemaphores    = &frame.present_semaphore;
+
+	VkResult result = vkQueuePresentKHR(queue->get_handle(), &present);
+
+	add_present_to_history(index);
+	cleanup_present_history();
+
+	return result;
+}
+
+void SwapchainRecreation::add_present_to_history(uint32_t index)
+{
+	PerFrame &frame = submit_history[submit_history_index];
+
+	present_history.emplace_back();
+	present_history.back().present_semaphore = frame.present_semaphore;
+	present_history.back().old_swapchains    = std::move(old_swapchains);
+
+	frame.present_semaphore = VK_NULL_HANDLE;
+
+	// The fence needed to know when the semaphore can be recycled will be one that is passed to
+	// vkAcquireNextImageKHR that returns the same image index.  That is why the image index
+	// needs to be tracked in this case.
+	present_history.back().image_index = index;
+}
+
+void SwapchainRecreation::cleanup_present_history()
+{
+	while (!present_history.empty())
+	{
+		PresentOperation &present_operation = present_history.front();
+
+		// If there is no fence associated with the history, it can't be cleaned up yet.
+		if (present_operation.cleanup_fence == VK_NULL_HANDLE)
+		{
+			// Can't have an old present operation without a fence that doesn't have an
+			// image index used to later associate a fence with it.
+			assert(present_operation.image_index != INVALID_IMAGE_INDEX);
+			break;
+		}
+
+		// Otherwise check to see if the fence is signaled.
+		VkResult result = vkGetFenceStatus(get_device_handle(), present_operation.cleanup_fence);
+		if (result == VK_NOT_READY)
+		{
+			// Not yet
+			break;
+		}
+		VK_CHECK(result);
+
+		cleanup_present_operation(present_operation);
+		present_history.pop_front();
+	}
+
+	// The present history can grow indefinitely if a present operation is done on an index
+	// that's never acquired in the future.  In that case, there's no fence associated with that
+	// present operation.  Move the offending entry to last, so the resources associated with
+	// the rest of the present operations can be duly freed.
```

Comment:
Simple :)

Step 1: Write a driver that does this
Step 2: Run this sample

But seriously, it's out of our control. If you really want to provoke this, you could probably modify the swapchain code in mesa to create 1 additional image, then only return that image on like the 5th acquire and never again.

As to why I'm guarding against this, it's because this might actually happen in practice. Like if you create a MAILBOX swapchain with 3 images, then have a burst of frames that makes the swapchain return all images, then really slow down and cycle through images 0 and 1, then image 3 may very well never be returned again by the swapchain implementation. You could imagine the same with FIFO for that matter.